torch/csrc/jit/interpreter.cpp

```diff
@@ -627,21 +628,48 @@ struct CodeImpl {
   }
 
   void emitBailOut(Node* node) {
-    // BailOut node has the `attr::Subgraph` which
-    // contains the original deoptimized version
-    // of a computational graph starting from the
-    // bailout point.
-    // BailOut node's first input is a guarded tensor
-    // the rest are inputs we need to be able to
-    // execute the bailout graph
-    emitLoadInputs(node->inputs().slice(0, 1));
+    auto unoptimized_graph = node->inputs()
+                                 .at(0)
+                                 ->type()
+                                 ->expect<FunctionType>()
+                                 ->function()
+                                 ->graph();
+    // unoptimized graph is at index 0
+    // guarded input is at index 1
+    // the rest of args follow
+    emitLoadInputs(node->inputs().slice(1, 1));
     insertInstruction(GUARD, type_table_.size());
     type_table_.emplace_back(node->outputs().at(0)->type());
     insertInstruction(JF, 0 /* to be patched */);
+
     size_t jf_index = instructions_.size() - 1;
-    emitLoadInputs(node->inputs().slice(1));
+    // note, guaded input is already loaded onto the stack
+    // for GUARD instruction
+    emitLoadInputs(node->inputs().slice(2));
     insertInstruction(TAIL_CALL, function_table_.size());
-    auto func = std::make_shared<Function>("bailout", /*optimize=*/true, node->g(attr::Subgraph), nullptr);
+    TORCH_INTERNAL_ASSERT(node->kind() == prim::BailOut);
+    auto bailout_index = node->i(attr::index);
+    TORCH_INTERNAL_ASSERT(bailout_index >= 0);
+
+    auto build_bailout_graph = [bailout_index,
+                                unoptimized_graph](Function& func) {
+      auto bailout_graph =
+          BuildBailOutGraphFrom(bailout_index, unoptimized_graph);
+      auto bgb = bailout_graph->block();
+      std::unordered_map<Value*, Value*> uncaptures_map;
+      auto env = [&bgb, &uncaptures_map](Value* v) -> Value* {
+        if (!uncaptures_map.count(v)) {
+          // Capture values for both graphs
+          uncaptures_map[v] = bgb->addInput()->copyMetadata(v);
+        }
+        return uncaptures_map[v];
+      };
+      func.graph()->block()->cloneFrom(bailout_graph->block(), env);
```

Comment:
This is weird. Can we just have BuildBailOutGraph take the functions graph to fill in? This is an extra graph copy here and it would remove the capture logic right?

Reply:
yup